src/Symfony/Component/Lock/Store/MysqlStore.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock\Store;
+
+use Symfony\Component\Lock\Exception\InvalidArgumentException;
+use Symfony\Component\Lock\Exception\LockAcquiringException;
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\PersistingStoreInterface;
+
+/**
+ * @author rtek
+ * @author Jérôme TAMARELLE <jerome@tamarelle.net>
+ */
+class MysqlStore implements PersistingStoreInterface
+{
+    private \PDO $conn;
+
+    public function __construct(\PDO $conn)
+    {
+        if ('mysql' !== $driver = $conn->getAttribute(\PDO::ATTR_DRIVER_NAME)) {
+            throw new InvalidArgumentException(sprintf('The adapter "%s" does not support the "%s" driver.', __CLASS__, $driver));
+        }
+
+        if (\PDO::ERRMODE_EXCEPTION !== $conn->getAttribute(\PDO::ATTR_ERRMODE)) {
+            throw new InvalidArgumentException(sprintf('"%s" requires PDO error mode attribute be set to throw Exceptions (i.e. $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION)).', __METHOD__));
+        }
+
+        $this->conn = $conn;
+    }
+
+    public function save(Key $key): void
+    {
+        if ($key->hasState(__CLASS__)) {
+            return;
+        }
+
+        // todo ? check that mysql > 5.7.3
+
+        // mysql limits lock name length to 64 chars
+        $name = (string) $key;
+        $name = \strlen($name) > 64 ? hash('sha256', $name) : $name;
+
+        $stmt = $this->conn->prepare('SELECT IF(IS_USED_LOCK(:name) = CONNECTION_ID(), -1, GET_LOCK(:name, 0))');
```

Comment:
`PostgreSqlStore` uses the InMemoryStore to avoids duplicate lock on the same connection. 
https://github.com/symfony/symfony/blob/1f514f97c0dd3c12fbe47787dae17b6e56e2637f/src/Symfony/Component/Lock/Store/PostgreSqlStore.php#L288

Reply:
Can you elaborate on this requirement?

Im concerned that using the InMemoryStore can cause unexpected behavior in situations where either:
1) A combined store is used with with Mysql and Postgres
2) A the same key is passed to two `LockFactory::createLockFromKey`, one using mysql and the other using postgres

I moved the state into the store since the tests state that `exists()` is bound to they `Key` instance, while `save()` is bound to the `Key::$resource`. Does this work?

Reply:
1. They should not share the same instance of `InMemoryStore`